### ISSUE TITLE
Add validationSchema to FormikProps

### ISSUE THREAD

**Issue by klis87** [jaredpalmer/formik#797]
## Current Behavior
One of the reasons Formik is so great is that it plays very nicely with `yup`. With `yup`, we dont need to combine validations with ui, we can use `yup` for validations and Formik just for view layer. Especially for complex forms, it really simplifies the code, makes it more understandable and more testable.

However, there is a small price to pay, sometimes ui might need the access to validation rules, for example to display `*` for required fields. Right know the answer is to use an additional prop passed to `Field` component, like `required`. Unfortunately this is a duplication, as in this case we already used `yup.string().required()` for instance. 

## Desired Behavior

With `validationSchema` passed as part of `FormikProps`, schema itself could be used to get required information.

## Suggested Solution

If he had `schemaValidation` object passed to `Field` component, we could use `yup` `reach` to get to schema part we are interested in a given field and `describe` to get tests for the give schema, like `required`

## Who does this impact? Who is this for?

It gives an additional, but optional option to everyone to style field components.

## Describe alternatives you've considered

Code duplication.

## Additional context

N/A

**Comment by melkir** [comment]
This may be related to https://github.com/jaredpalmer/formik/commit/739d7e1e7e2dd66cf76e9410b313a3b2254b3b6d
We no longer have access to the props here 
https://github.com/jaredpalmer/formik/blob/master/src/Formik.tsx#L249

**Comment by klis87** [comment]
@melkir No, this is something different - this is about creating dynamic schema based on props, this topic is about accessing current schema which Formik uses for validation from `Field` component

**Comment by klis87** [comment]
@jaredpalmer Writing again not to be autoclosed by the bot. Could you please tell me whether do you like this idea or you need any more info? I really need this feature for `required` state and I really don't wanna describe this value both in Formik components and Yup schemas.

If you like the idea, I could work on PR for this.

**Comment by cjones26** [comment]
FYI -- I accomplished this like so:

```
export default class DashboardView extends Component {
  constructor(props) {
    super(props);

    this.state = {
      initialModel: {
        textInputTest: ''
      }
    };

    this.validationSchema = Yup.object().shape({
      textInputTest: Yup.string().required('textInputTest is required')
    });
  }

  isRequired = path => {
    try {
      return (
        Yup.reach(this.validationSchema, path)
          .describe()
          .tests.indexOf('required') > -1
      );
    } catch (error) {
      return false;
    }
  };

  render() {
    const { initialModel, validationSchema } = this.state;

    return (
      <div>
        <h1>Dashboard</h1>
        <Formik
          validationSchema={validationSchema}
          initialValues={initialModel}
          onSubmit={(values, actions) => {
            console.log('values: ', values);
            console.log('actions: ', actions);
          }}
          render={({ errors }) => (
            <Form>
              <TextInputWrapped name="textInputTest" label="Text Input" required={this.isRequired('textInputTest')} />
 
              <Button type="submit" kind="primary" disabled={errors.length > 0}>
                Submit
              </Button>
            </Form>
          )}
        />
      </div>
    );
  }
}
```

**Comment by klis87** [comment]
@cjones26 The problem is that I dont want to, or even in some cases cannot pass any props to my custom fields component, as they are often nested, and in some situation I create my forms dynamically, for instance by WYSIWIG and connect all pieces together with https://github.com/ianstormtaylor/slate/tree/master/packages/slate-html-serializer

React context is the way to go in this scenario, if we added `validationSchema` to Formik context, then this logic could be implemented on the field level, which is one of things so great about Formik, I can keep my forms very DRY, required thing is the only one I cannot accomplish, which also is quite a common scenario in many apps to show some required field indicator like `*`

**Comment by klis87** [comment]
@cjones26 no worries, and I appreciate it :) thx for your code snippet, your solution is exactly the same as I imagined, you just passed `required` as prop, not using React context, but nevertheless this is a proof that this is not a feature which would be used only by me :) so, @jaredpalmer , I really can work on this, I just need a green light that you won't rejected PR with this idea :)

**Comment by thomasmery** [comment]
Hi there,

just like to chime in to say I was also looking for the Form.validationSchema in the form prop passed to custom component 

thanks @cjones26 & @klis87 

@jaredpalmer if you have time to let us know whether this could be considered

thanks a lot

**Comment by todda00** [comment]
I also could benefit from having the validationSchema in the FormikBag, I too use yup.required() as the OP stated, and also use .label() to provide my form labels automatically. Currently I am providing my schema to my own components which add the label, required, etc.

```
const TextField = withProps({ schema })(Text);
...
<TextField name="title" />
```

Text component
```
const Text = ({
  name,
  label,
  error,
  required,
  field,
  form,
  type = "text",
  description,
  afterChange = () => null,
  ...props
}) => (
  <FieldWrap>
    {label && <Label required={required}>{label}</Label>}
    <StyledTextInput
      id={name}
      type={type}
      {...field}
      onChange={(...changeProps) => {
        form.handleChange(...changeProps);
        const [event] = changeProps;
        afterChange(event.target.value, { field, form });
      }}
      {...props}
    />
    {description ? <FieldDescription>{description}</FieldDescription> : null}
    {error ? <FieldError>{error}</FieldError> : null}
  </FieldWrap>
);

export default withFormikField(Text);
```

withFormikField
```
const withFormikField = Component => {
  const EnhancedComponent = withFieldProps(Component);
  const FormikFieldHOC = ({ fastField, ...rest }) => {
    const FormikField = fastField === true ? FastField : Field;
    return <FormikField {...rest} component={EnhancedComponent} />;
  };

  return FormikFieldHOC;
};
```

withFieldProps
```
const withFieldProps = withProps(
  ({
    name: propName,
    field, // { name, value, onChange, onBlur }
    form, // also values, setXXXX, handleXXXX, dirty, isValid, status, etc.
    label,
    schema,
    ...rest
  }) => {
    const { name: fieldName } = field || {};
    const { touched, errors } = form || {};
    const name = propName || fieldName;
    let fieldLabel = label;
    let tests = [];
    let schemaLabel;
    if (schema) {
      ({ label: schemaLabel, tests } = yup.reach(schema, name).describe());
    }

    if (fieldLabel === undefined) fieldLabel = schemaLabel;
    const touch = _.get(touched, name, null);
    const error = touch ? _.get(errors, name, null) : null;
    const required = tests.indexOf("required") !== -1;

    return {
      label: fieldLabel,
      error,
      name,
      required,
      field,
      form
    };
  }
);
```

If the validationSchema was passed in the form prop, I wouldn't have to wrap each of my different field types with `const TextField = withProps({ schema })(Text);` on every use, I could get to the schema from the reused field components.

**Comment by klis87** [comment]
@todda00 nice idea with `label`, now I want this feature even more :)

**Comment by klis87** [comment]
Anticloser message

**Comment by klis87** [comment]
@jaredpalmer what do u mean?

**Comment by andrewphillipoef** [comment]
I too expect that things like Required and Label being pulled out by formik would be excellent, is there a reason not to allow us to do this?

**Comment by DarrenAllen** [comment]
> s there a reason not to allow us to do this?
Ditto...

**Comment by knrz** [comment]
This would be really helpful. It's a 1-line code change (add validationSchema to the ctx object on line 823 of Formik.tsx). @jaredpalmer what is the reason for not passing in validationSchema?

**Comment by johnrom** [comment]
I think not passing through the schema helps prevent re-renders for users who do something like the following:

```
const MyForm = () => {
    const myValidationSchema = Yup.shape().object({
        name: Yup.string().required("Please enter your name.");
    });
    // this is a new object on every render, so 
    // anything passed validationSchema would re-render each time Formik re-renders.
    <Formik validationSchema={myValidationSchema} />
};
```

I'm not saying this is the right solution, but giving an example of why this might break Formik for users who have already implemented something like this.

**Comment by sanstream** [comment]
@cjones26 and @jaredpalmer Just ran into the same problem described above. being able to access the validation rules on a Field level to figure out if a field should be required or not can be very useful.

**Comment by mrmeek** [comment]
I would also like to be able to pull the validation schema from the formik bag for the reasons other users have outlined above.

**Comment by johnrom** [comment]
Not stale
